File: JS/NumberOfPeopleAwareOfSecret.js
/**
 * @version JAVASCRIPT ECMAScript 6
 * @author MARIO SAM <eu@mariosam.com.br>
 * @see I would love to work with you instead solving web code tests: hire me!
 */

/**
 * @param {number} n
 * @param {number} delay
 * @param {number} forget
 * @return {number}
 */
function peopleAwareOfSecret(n, delay, forget) {
    const dp = Array(n + 1).fill(0)
    const mod = 1e9 + 7
    
    dp[1] = 1
    for(let i = 1; i <= n; i++) {
        for(let j = i + delay; j < i + forget; j++) {
        if(j > n) break
        dp[j] += dp[i]
        dp[j] %= mod
        }
    }

    let res = 0
    for(let i = n - forget + 1; i <= n; i++) {
        res = (dp[i] + res) % mod
    }
    
    return res
}
export { peopleAwareOfSecret }

File: JAVA/NumberOfPeopleAwareOfSecret.java
/**
 * @version JAVA
 * @author MARIO SAM <eu@mariosam.com.br>
 * @see I would love to work with you instead solving web code tests: hire me!
 */
package JAVA;

public class NumberOfPeopleAwareOfSecret {

    public static void main(String[] args) {
        System.out.printf("Resultado: %d\n", peopleAwareOfSecret( 2,3,4 ));
    }

    public static int peopleAwareOfSecret(int n, int delay, int forget) {
        int mod = 1_000_000_007;
        long[] dp = new long[n + 1];
        dp[1] = 1;

        for (int i = 1; i <= n; i++) {
            for (int j = i + delay; j < i + forget && j <= n; j++) {
                dp[j] = (dp[j] + dp[i]) % mod;
            }
        }

        long res = 0;
        for (int i = n - forget + 1; i <= n; i++) {
            res = (res + dp[i]) % mod;
        }

        return (int) res;
    }

}
